datafusion/physical-plan/src/joins/sort_merge_join.rs

```diff
@@ -860,6 +867,8 @@ impl SMJStream {
                                 self.reservation.shrink(buffered_batch.size_estimation);
                             }
                         } else {
+                            // If the head batch is not fully processed, break the loop.
+                            // Streamed batch will be joined with the head batch in the next step.
```

Comment:
Not related to this issue. But as I re-read these code, I add some comments to help us understand it again.